### DISCUSSION TITLE
Is SSR needed in main?

### DISCUSSION THREAD

**Discussion by mbforbes** [cawa-93/vite-electron-builder#1003] [Q&A] [answered]
I'm hoping to create a minimal config using this repository (thank you!) as a baseline. I'm trying to understand the effect of SSR in Vite's config in the main package.

Referencing [these couple lines](https://github.com/cawa-93/vite-electron-builder/blob/main/packages/main/vite.config.js#L21-L23):

```
    ssr: true,
    sourcemap: 'inline',
    target: `node${node}`,
```

I was surprised that just setting `target: "node20"`, Vite would still build the main package for the browser. (And electron would fail to run it.) But setting `ssr: true` seems to force Vite to build main for node. I read through https://github.com/cawa-93/vite-electron-builder/issues/655 but still don't fully understand.

So my questions are:
- is `target` working as intended? or is it not actually doing anything?
- do we really need SSR? is it only to get a node build?

I guess I feel worried setting SSR just to get a node build, because my app doesn't actually use SSR. But I'm very new to electron so I'm probably missing a lot.

Thank you for any insights! 🙏

**Comment by cawa-93** [comment, marked as answer]
- The `target` property used by transpires. It defines what language features can be used and for what polyfill is needed, what CSS properties need vendor prefix, etc.
- `ssr: true`  define a javascript runtime. 

In context of `main` module, term `ssr` doesn't mean anything because main doesn't render anything. And this option don't bring anithind extra in runtime. It is just a easy way to tell Vite to build some code and do not scary internal `node` packages.

**Reply by mbforbes** [reply]
Thank you for the answer. It's been enough months now that I don't fully understand my question or your answer 😄 

I think what was confusing me is that
- `ssr: false` builds `main` for the browser (which never makes sense to do), and
- setting `ssr: true`, builds `main` for node, which is always what we want, and is unrelated to server-side rendering

To add to my confusion, `target` _seemed_ to say "are we building this for the browser or node?" but wasn't actually doing that.

My understanding of your answer is that:
- `target` — does not define the execution platform to build for (e.g., browser vs node). rather, it specifically describes lower-level ECMAScript language features
- `ssr` — is currently used as a way to set the execution platform to node. when we set this, vite will correctly import all the node packages, allowing our main to be run by electron

is this correct?

**Reply by cawa-93** [reply]
Correct

**Comment by erikjalevik** [comment]
I had the same question, and I found that another thing `ssr: true` seems to cause is automatic externalising of all dependencies. I happened to want them bundled inside main.js, and the only way I could find of doing that was to set `ssr: false`.
